chain-abci/src/app/app_init/validator_state.rs

```diff
@@ -35,12 +35,180 @@ pub struct ValidatorState {
+type DeleteSchedule = (StakedStateAddress, (Timespec, TendermintValidatorAddress));
+
+impl Encode for ValidatorState {
+    fn size_hint(&self) -> usize {
+        let mut unbond_schedule_iter = self.after_unbond_delete.iter();
+        let ub_size = unbond_schedule_iter.len();
+        let ub_item_size = if ub_size > 0 {
+            unbond_schedule_iter.next().size_hint()
+        } else {
+            0
+        };
+
+        self.council_nodes_by_power.size_hint()
+            + self.tendermint_validator_addresses.size_hint()
+            + self.punishment.size_hint()
+            + self.proposer_stats.size_hint()
+            + ub_size * ub_item_size
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        self.council_nodes_by_power.encode_to(dest);
+        self.tendermint_validator_addresses.encode_to(dest);
+        self.punishment.encode_to(dest);
+        self.proposer_stats.encode_to(dest);
+        let unbond_schedule: Vec<DeleteSchedule> = self
+            .after_unbond_delete
+            .iter()
+            .map(|x| (*x.0, x.1.clone()))
+            .collect();
+        unbond_schedule.encode_to(dest);
+    }
+}
+
+impl Decode for ValidatorState {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+        let council_nodes_by_power =
+            BTreeMap::<(TendermintVotePower, StakedStateAddress), CouncilNode>::decode(input)?;
+        let tendermint_validator_addresses =
+            BTreeMap::<TendermintValidatorAddress, StakedStateAddress>::decode(input)?;
+        let punishment = ValidatorPunishment::decode(input)?;
+        let proposer_stats = BTreeMap::<StakedStateAddress, u64>::decode(input)?;
+        let unbond_schedule: Vec<DeleteSchedule> = Vec::decode(input)?;
+        let after_unbond_delete: DeleteScheduleMap = IndexMap::from_iter(unbond_schedule);
+        let mut validator_state_helper = ValidatorStateHelper::default();
+        for key in council_nodes_by_power.keys() {
+            validator_state_helper
+                .validator_voting_power
+                .insert(key.1, key.0);
+        }
+        Ok(ValidatorState {
+            council_nodes_by_power,
+            tendermint_validator_addresses,
+            punishment,
+            proposer_stats,
+            after_unbond_delete,
+            validator_state_helper,
+        })
+    }
+}
 
 impl ValidatorState {
+    pub fn lowest_vote_power(&self) -> TendermintVotePower {
+        self.council_nodes_by_power
+            .keys()
+            .next()
+            .expect("at least one validator")
+            .0
+    }
+
+    /// number of current validators
+    pub fn number_validators(&self) -> usize {
+        self.council_nodes_by_power.len()
+    }
+
+    /// Removal of council node metadata at the begin block
+    pub fn metadata_clean(&mut self, current_time: Timespec) {
+        match self.after_unbond_delete.get_index(0) {
```

Comment:
The business logic requires iterating `after_unbond_delete` by order of time, `IndexMap` keeps the order of insertion, so it's only correct when the order of insertion and order of time are the same.

Reply:
I can put an extra comment about this assumption -- so far it works, because the insertion order = block order and the time in blocks should be monotonic + it's using `shift_remove` instead of `remove` (that would scramble the order)

Reply:
The more stable approach might need to use two maps, one is indexed by address for querying, one is indexed by time for iterating. Encapsulate the operations to ensure data consistency.

Reply:
That may be better / less error-prone, thanks!
I'll first add a few more tests on the current logic -- as cleanups don't seem to be covered (the multi-node unjail->nodejoin integration test works now though): https://codecov.io/gh/crypto-com/chain/pull/1008/changes#D1-121 
and then change this `after_unbond_delete` queue to that